lib/matplotlib/figure.py

```diff
@@ -964,13 +965,44 @@ def set_facecolor(self, color):
         """
         self.patch.set_facecolor(color)
 
-    def set_dpi(self, val):
+    def set_dpi(self, dpi):
         """
-        Set the resolution of the figure in dots-per-inch.
+        Set the nominal resolution of the figure in dots-per-inch.
 
-        .. ACCEPTS: float
-        """
-        self.dpi = val
+        This is nominal because some screens have "hi-dpi" or "Retina",
+        where pixels are doubled.  This dpi is the non-doubled value.
+
+        Setting the dpi different from your screen dpi makes the figure larger
+        or smaller on the screen than the width and height specified in
+        `~.Figure.get_width_height`.
+
+        Parameters
+        ----------
+        dpi : float
+
+        Notes
+        -----
+        Some backends have the concept of "hi-dpi" displays (or "Retina")
+        where the resolution is doubled, but dimensions are traditionally
+        specified as if the resolution were not doubled.  Some Matplotlib
+        GUI backends respect this doubling (i.e. Qt5Agg), but Matplotlib still
+        specifies dimensions and dpi to matplotlib as if the resolution
+        were not doubled. The default figure
+        value is nominally 100 dpi, but displays that are deignated "hi-dpi"
+        will internally double this to 200 dpi.  We keep the "nominal" dpi
+        because some computer set ups have one display at normal dpi, and a
+        second at hi-dpi, so a nominal resolution must be stored to stop
+        figures from doubling or halving is size when moved between displays.
+        """
+
+        # some backends set self._dpi to be a ratio times
+        # self._original_dpi:
+        ratio = self._dpi / self._original_dpi
```

Comment:
I am not sure that we want to rely on this here.  I can not produce one off the top of my head, but this feels like it is prone to race conditions.

This dpi doubling is the business of the GUI backends and I would like to keep as much of it there as possible.  What we should do is set the `_original_dpi` and then let the GUI know and have it sort the doubling out.

Reply:
Oh, no its not that simple because of interactivity.  I think it really is better to actually set the dpi for the figure object so it knows its actual rendered dpi....